argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+			dec := json.NewDecoder(stdoutPipe)
+			dec.DisallowUnknownFields()
+			var rsp *Argon2OutOfProcessResponse
+			if err := dec.Decode(&rsp); err != nil {
+				return fmt.Errorf("cannot decode response: %w", err)
+			}
+
+			if rsp.Err() != nil {
+				// If we receive an error response, begin the process of the tomb dying.
+				// Don't wrap the error - this will be returned directly to the caller.
+				return rsp.Err()
+			}
+
+			switch rsp.Command {
+			case Argon2OutOfProcessCommandWatchdog:
+				// Direct watchdog responses to wdRspChan so they can be picked up by
+				// the watchdog handler.
+				select {
+				case wdRspChan <- rsp: // Unbuffered channel, but read end is always there unless the tomb is dying.
+				case <-tmb.Dying():
+					// The loop will no longer iterate
+				}
+			default:
+				// For any other response, first of all make sure that the command value is
+				// consistent with the sent command.
+				if rsp.Command != req.Command {
+					// Unexpected command. Return an appropriate error to begin the process
+					// of the tomb dying
+					return &Argon2OutOfProcessResponseCommandInvalidError{
+						Response: rsp.Command,
+						Expected: req.Command,
+					}
+				}
+				// If it is consistent, save the response to return to the caller and begin a clean
+				// shutdown of the tomb.
+				actualRsp = rsp
+				tmb.Kill(nil)
+				// This loop will no longer iterate
+			}
+		}
+		return tomb.ErrDying
+	})
+
+	// Wait here until the tomb enters a dying state
+	<-tmb.Dying()
+
+	// Closing the stdin pipe is necessary to unblock WaitForAndRunArgon2OutOfProcessRequest
+	// on the remote side, if it is blocked in a read.
+	if err := stdinPipe.Close(); err != nil {
+		return nil, fmt.Errorf("cannot close stdin pipe: %w", err)
+	}
+	// The go documentation says this should be unnecessary because it will be closed
+	// by cmd.Wait once the command has exitted. But it's still possible for
+	// WaitForAndRunArgon2OutOfProcessRequest to be blocked on a write to us, which
+	// will never complete once that the loop that handles responses has terminated.
+	// Therefore, we're going to close it explicitly.
+	if err := stdoutPipe.Close(); err != nil {
+		return nil, fmt.Errorf("cannot close stdout pipe: %w", err)
+	}
+
+	// We can wait for the remote process to exit now.
+	if err := cmd.Wait(); err != nil {
+		return nil, fmt.Errorf("an error occurred whilst waiting for the remote process to finish: %w", err)
+	}
+	// Stop the 5 second kill timer
+	close(exitWaitCh)
+
+	// Wait for all go routines to finish.
+	if err := tmb.Wait(); err != nil {
+		// Don't wrap this error - this will be the first non-nil error passed
+		// to Tomb.Kill. There's no benefit to adding additional context here.
+		return nil, err
+	}
+
+	return actualRsp, nil
+}
+
+func (k *outOfProcessArgon2KDFImpl) Derive(passphrase string, salt []byte, mode Argon2Mode, params *Argon2CostParams, keyLen uint32) (key []byte, err error) {
```

Comment:
I know it's part of the existing interface, but if params is always expected to be non nil, this could pass value not a pointer.

Reply:
That seems reasonable, but given that this interface has been around for a while, shall we leave this one for another PR?